### PR TITLE
Fix outdated & upgrade commands

### LINKED ISSUE #111: `outdated` & `upgrade` fail on Catalina
Running on macOS High Siera Version 10.13.1 (17B1003), `mas outdated` doesn't print anything, even if there is an update for macOS 10.13.2 and I have the `App Store` opened.

### PR DESCRIPTION
**Pull request by rgoldberg**
Fix outdated & upgrade commands.

Resolves #111 #245 #248 #252 #282 #306

Possibly #76 #92 #247 #249 #290

### CODE REVIEW

**Comment by rgoldberg** [comment]
I force pushed some minor changes 14 hours ago, but the Jenkins/pr-merge check hasn't completed.  How long does it normally take?  Might something have gone wrong?  14 hours seems a long time, but maybe it's normal…

**Comment by rgoldberg** [comment]
The most recent Jenkins/pr-merge build failed, but it previously succeeded.

The logs say that the Commandant build failed.

The only differences in the most recent force push were in `CHANGELOG.md`, which should not affect the Commandant build, right?

Supposedly, there's more info in the following file, but I don't know how to see the file:

```
/var/folders/8l/wsf0fx0x015dtgb_pwjyj3zr0000gr/T/carthage-xcodebuild.RDNixk.log
```

**Comment by rgoldberg** [comment]
Whenever I build, changes are made to spacing in various files that I haven't edited (presumably because the spacing violates various formatting rules).

Can I commit those spacing changes, so people don't keep getting files changed on them?

There's also a bug in the `InfoCommandSpec` `displays app details` check: the expected value for Released is `Jan 7, 2019`, but on my machine, the value is `2019-01-07`.  Either the default medium date style must be changed locally in the test function, or the system's default medium date style must be used to calculate the expected output.

**Comment by rgoldberg** [comment]
The file spacing is no longer changed by the build, presumably because of the change from `swiftlint autocorrect --format` to `swiftlint autocorrect`, but the `Info_command__displays_app_details` date format error still occurs.

**Comment by rgoldberg** [comment]
Hi, @phatblat, glad you’re recovering. Sorry to bother you, but is there anything that I (or anyone else) must do to get the `Jenkins/pr-merge` check to run? I pushed commits 2 days ago, and it still hasn’t run…

**Comment by phatblat** [comment]
@rgoldberg sorry for the delay getting to this PR. The Jenkins server these checks run on is a mac mini I have hosted at MacStadium. It's been offline for a while due to an OS upgrade that failed.

The CI just runs [several of the scripts](https://github.com/mas-cli/mas/blob/master/Jenkinsfile#L40) checked into the project, so as long as these are working for you and me, we can ignore Jenkins for now.

I will take a closer look at these changes and test them out myself tonight, but at a quick glance, I think this is a great idea. This tool has been using a private CommerceKit API to figure out which apps have pending updates but that API stopped returning anything some time back. I've dug around in a debugger several times looking for where this functionality moved to but haven't been successful.

Thanks for the PR! I will get back with any suggestions this week.

**Comment by rgoldberg** [comment]
Thanks, @phatblat. Glad you're recovering. No problem about delay. Take whatever time you need to stay healthy, and to ensure the PR is good.

I haven't looked throughout mas to see whether there are other commands that can be similarly fixed.

I tried to write tests, but got cryptic errors, so I decided to wait on tests until I communicated with you.

**Comment by phatblat** [comment]
Sorry about the changing format. I'd rather have that as a separate script but it was easy to add the --format flag to swiftlint and didn't think about the ramifications.

On the "InfoCommandSpec displays app details" test failures, I think a deterministic date format (option 2) would be ideal
since the output could be used by other CLI tools.

I'm curious, what locale setting do you have to get that format? That "yyyy-MM-dd" format (ISO-8601 minus the time) is perfect for the new format. You can find this formatter in [AppInfoFormatter](https://github.com/mas-cli/mas/blob/master/MasKit/Formatters/AppInfoFormatter.swift#L52).

**Comment by rgoldberg** [comment]
No problem about formatting.  I just didn't want to cause build or git history issues by changing things without asking.  I thought that maybe I had missed some setup instructions that would have prevented the file reformatting.

I set all 4 Date formats in System Preferences > Language & Region > Advanced… > Dates to "yyyy-MM-dd" a long time ago.

I'll have to read about the Apple `DateFormatter` to learn about how to set it programmatically, as I don't normally code in Swift or for Apple APIs.

Should I open a new issue & PR for the "InfoCommandSpec displays app details" date format change, or just include it in this PR in a separate commit?

**Comment by phatblat** [comment]
Yes, please create a separate PR for this date formatting change. That way it be easier to track see which version it changed in.

**Comment by rgoldberg** [comment]
I'll create a new issue & PR for the date format change after we finish with the current PR, as I don't want to have to manage multiple branches. The second PR will be simple, so easy to finish after this one.

**Comment by phatblat** [comment]
I managed to get the CI server back online tonight but builds are failing due to some issues I still need to clean up. So ignore the failing check for now.

**Comment by rgoldberg** [comment]
Will do. Please let me know whenever I need to do anything to improve / finish the PR.

**Comment by phatblat** [comment]
Wow, this works better than the App Store's "Updates" section! The App Store app doesn't show any of the 5 pending updates.

```
↪ mas outdated
1462114288 Grammarly for Safari (9.27 -> 9.26)
803453959 Slack (4.12.2 -> 4.13.0)
Warning: Identifier 927292435 not found in store. Was expected to identify iStat Mini.
411643860 DaisyDisk (4.12.1 -> 4.12)
1475897096 Jira (122.0.0 -> 123.1.0)
Warning: Identifier 781577745 not found in store. Was expected to identify Dumper.
Warning: Identifier 924726344 not found in store. Was expected to identify Deliveries.
1246284741 Install macOS High Sierra (13.6.02 -> 10.13.6)
```

Looked into the apps which weren't found:

- iStat Mini - no longer in app store
- Dumper -  no longer in app store
- Deliveries - has a different app store identifier now: 290986013

There might be a way to map apps like Deliveries that release updates as a new app to the one, but these are technically not an upgrade so the only thing we might be able to do is to have a more informative warning message.

The macOS installers have a weird versioning scheme which looks like it doesn't match the version reported by the iTunes Search API. This is a very minor issue as these are not usually kept around.

**Comment by phatblat** [comment]
Weird, both Grammarly and DaisyDisk are newer versions on my system than the app store reports.

- 1462114288 Grammarly for Safari (9.27 -> 9.26)
- 411643860 DaisyDisk (4.12.1 -> 4.12)

I installed them from the app store some time ago. Not sure how this can happen unless installing beta versions direct from the developer (which these aren't). I'll keep digging on the reason for this.

**Comment by phatblat** [comment]
Running the `upgrade` command attempted to upgrade all 8 of the items listed by the `outdated` command, even the ones that output a warning.

```
↪ mas upgrade
Upgrading 8 outdated applications:
Grammarly for Safari (9.27), Slack (4.12.2), iStat Mini (1.1), DaisyDisk (4.12.1), Jira (122.0.0), Dumper (1.2.1), Deliveries (3.2.4), Install macOS High Sierra (13.6.02)
==> Downloading Grammarly for Safari
==> Installed Grammarly for Safari
==> Downloading Slack
==> Installed Slack
==> Downloading iStat Mini
==> Installed iStat Mini
==> Downloading DaisyDisk
==> Installed DaisyDisk
==> Downloading Jira Cloud by Atlassian
==> Installed Jira Cloud by Atlassian
==> Downloading Dumper
==> Installed Dumper
==> Downloading Deliveries 3
==> Installed Deliveries 3
==> Downloading macOS High Sierra
==> Installed macOS High Sierra
```

Afterwards `outdated` lists 6 apps. Slack and Jira were the only two successfully upgraded.

```
↪ mas outdated
1462114288 Grammarly for Safari (9.27 -> 9.26)
Warning: Identifier 927292435 not found in store. Was expected to identify iStat Mini.
411643860 DaisyDisk (4.12.1 -> 4.12)
Warning: Identifier 781577745 not found in store. Was expected to identify Dumper.
Warning: Identifier 924726344 not found in store. Was expected to identify Deliveries.
1246284741 Install macOS High Sierra (13.6.02 -> 10.13.6)
```

Is there a way to exclude the items that generate a warning from the `upgrade` command?

File: MasKit/Commands/Upgrade.swift [hunk 1]
```diff
@@ -16,71 +15,78 @@ public struct UpgradeCommand: CommandProtocol {
...
+                                // if argument a UInt64, lookup app by id using argument
+                                return appLibrary.installedApp(forId: appId)
+                            } else {
+                                // if argument not a UInt64, lookup app by name using argument
+                                return appLibrary.installedApp(named: $0)
+                            }
+                        }
+                ).compactMap {
+                    // only upgrade apps whose local version differs from the store version
+                    try storeSearch.lookup(app: $0.itemIdentifier.intValue)?.version != $0.bundleVersion
```
Comment (phatblat) [inline MasKit/Commands/Upgrade.swift:51]: This should be split to return nil when the app isn't found in the store. That should exclude missing apps from upgrade.

**Review by phatblat — CHANGES_REQUESTED** [review]

**Comment by rgoldberg** [comment]
> Weird, both Grammarly and DaisyDisk are newer versions on my system than the app store reports.
> 
> I installed them from the app store some time ago. Not sure how this can happen unless installing beta versions direct from the developer (which these aren't). I'll keep digging on the reason for this.

Maybe a version was released to the store, but was later retracted…

File: MasKit/Commands/Upgrade.swift (hunk 1 above)
Comment (rgoldberg) [reply MasKit/Commands/Upgrade.swift:51]: Will get to this sometime over the weekend. Thanks.

Comment (rgoldberg) [reply MasKit/Commands/Upgrade.swift]: Sorry for delay.

I pushed a fix for this.

Please let me know if it works.

Will create an issue & PR for the date issue.

Hope hands / wrists are doing better.

**Comment by rgoldberg** [comment]
The `Jenkins/pr-merge` timed out, but I'm not sure why…

**Review by phatblat — APPROVED** [review]
Thanks for all your great work on this! 💯

**Comment by rgoldberg** [comment]
@phatblat No problem. Glad to help. Thanks for maintaining / working on mas.

All but one issue that I referenced in the commit message are still open, so someone should go through the list and close the applicable issues. Some issues will definitely be fixed by this PR, some other issues might be fixed, but I'm not 100% sure as I didn't have the time to verify each.

**Comment by phatblat** [comment]
Thanks for reminding me. Once I get #327 completed I'll ask on each issue for confirmation.

**Comment by rgoldberg** [comment]
The issues listed after `Resolves` should all be fixed by this PR.  Those listed after `Possibly` are ones I'm not 100% sure about.

**Pull request merged**